File: common/src/main/java/generations/gg/generations/core/generationscore/util/TaskQueue.java
package generations.gg.generations.core.generationscore.util;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

public class TaskQueue {
    private final BlockingQueue<Runnable> taskQueue;
    private final Thread workerThread;
    private final AtomicBoolean running;

    public TaskQueue() {
        taskQueue = new LinkedBlockingQueue<>();
        running = new AtomicBoolean(true);
        workerThread = new Thread(this::processTasks);
        workerThread.start();
    }

    private void processTasks() {
        while (true) {
            try {
                if (!taskQueue.isEmpty()) {
                    Runnable task = taskQueue.poll(100, TimeUnit.MILLISECONDS);
                    if(task != null) task.run();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    public void addTask(Runnable task) {
        taskQueue.offer(task);
    }

    public void start() {
        running.set(true);
    }

}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/util/TaskQueue.kt
package generations.gg.generations.core.generationscore.common.util

import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class TaskQueue {
    private val taskQueue: BlockingQueue<Runnable> = LinkedBlockingQueue()
    private val workerThread: Thread
    private val running = AtomicBoolean(true)

    init {
        workerThread = Thread { this.processTasks() }
        workerThread.start()
    }

    private fun processTasks() {
        while (true) {
            try {
                if (!taskQueue.isEmpty()) {
                    val task = taskQueue.poll(100, TimeUnit.MILLISECONDS)
                    if (task != null) {
                        try {
                            task.run()
                        } catch (e: Exception) {
                            println("Error!")
                            e.printStackTrace()
                        }
                    }
                }
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
                break
            }
        }
    }

    fun addTask(task: Runnable) {
        taskQueue.offer(task)
    }

    fun start() {
        running.set(true)
    }
}

